src/rust/generator-dispatcher/src/lib.rs

```diff
@@ -1 +1,223 @@
+                                                plugin_work_queue_client.clone(),
+                                                generator_ids,
+                                                envelope
+                                            ).await
+                                        } else {
+                                            // cache miss, but an update was
+                                            // successfully enqueued so we'll
+                                            // retry the message
+                                            retry_message(
+                                                raw_logs_retry_producer.clone(),
+                                                envelope
+                                            ).await?;
+                                            Ok(())
+                                        }
+                                    }
+                                    Err(cache_err) => {
+                                        if let GeneratorIdsCacheError::Retryable(reason) = cache_err {
+                                            tracing::warn!(
+                                                message = "generator IDs cache error, retrying message",
+                                                reason =% reason,
+                                            );
+                                            retry_message(
+                                                raw_logs_retry_producer.clone(),
+                                                envelope
+                                            ).await?;
+                                            Ok(())
+                                        } else {
+                                            // fatal error, bailing out
+                                            Err(GeneratorDispatcherError::from(cache_err))
+                                        }
+                                    }
+                                }
+                            }
+                            Err(e) => {
+                                tracing::error!(
+                                    message="error processing kafka message",
+                                    reason=%e,
+                                );
+
+                                Ok(())
+                            }
+                        }
+                    }
+                })
+                .buffer_unordered(pool_size);
+
+            pin_mut!(buffered);
+
+            while let Some(result) = buffered.next().await {
+                if let Err(e) = result {
+                    tracing::error!(
+                        message = "fatal error",
+                        reason =% e,
+                    );
+                    return Err(e);
+                }
+            }
+        }
+    }
+}
+
+async fn retry_message(
+    raw_logs_retry_producer: RetryProducer<RawLog>,
```

Comment:
i suspect you can just &RetryProducer here and save a clone, same for the clients, but nbd

Reply:
Cloning a producer is a cheap operation https://docs.rs/rdkafka/latest/rdkafka/producer/struct.FutureProducer.html

Reply:
You're right, it borrows `send(&self...` so this can be `&RetryProducer`.